Wangscape/tilegen/CartesianPowerIterator.h

```diff
@@ -0,0 +1,255 @@
+
+    void updateValue(size_t n);
+    void resetComponent(size_t n);
+    void incrementComponent(size_t n);
+};
+
+template<typename ForwardIt>
+inline CartesianPowerIterator<ForwardIt>::CartesianPowerIterator(ForwardIt first_, size_t base_size_, size_t init_coordinate_, size_t power_) :
+    mFirst(first_), mBaseSize(base_size_),
+    mCoordinates(power_, init_coordinate_)
+{
+    if (init_coordinate_ != getBaseSize())
+    {
+        ForwardIt init_iterator = getFirst();
+        std::advance(init_iterator, init_coordinate_);
+        mIterators.assign(power_, init_iterator);
+        mValues.assign(power_, *init_iterator);
+    }
+}
+
+template<typename ForwardIt>
+inline CartesianPowerIterator<ForwardIt>& CartesianPowerIterator<typename ForwardIt>::operator++()
+{
+    for (size_t i = 0; i < getPower(); i++)
+    {
+        size_t coordinate = getCoordinate(i);
+        if (coordinate >= getBaseSize())
+            break; // invalid or isEnd()
+        if (coordinate == getMaxCoordinate())
+        {
+            if (i == getMaxComponent())
+            {
+                ++mCoordinates[i];
+                break; // isEnd()
+            }
+            else
+            {
+                resetComponent(i);
+            }
+        }
+        else
+        {
+            incrementComponent(i);
+            break;
+        }
+    }
+    return *this;
+}
+
+template<typename ForwardIt>
+inline bool CartesianPowerIterator<ForwardIt>::isEnd() const
+{
+    return *getCoordinates().crbegin() >= getBaseSize();
+}
+
+template<typename ForwardIt>
+inline bool CartesianPowerIterator<ForwardIt>::operator==(const CartesianPowerIterator& other) const
+{
+    if (getFirst() != other.getFirst())
+        return false;
+    if (getBaseSize() != other.getBaseSize())
+        return false;
+    if (getPower() != other.getPower())
+        return false;
+    if (isEnd() && other.isEnd())
+        return true;
+    return getCoordinates() == other.getCoordinates();
+}
+
+template<typename ForwardIt>
+inline bool CartesianPowerIterator<ForwardIt>::operator!=(const CartesianPowerIterator& other) const
+{
+    return !(*this == other);
+}
+
+template<typename ForwardIt>
+inline typename CartesianPowerIterator<ForwardIt>::reference CartesianPowerIterator<ForwardIt>::operator*() const
+{
+    return getValues();
+}
+
+template<typename ForwardIt>
+inline typename CartesianPowerIterator<ForwardIt>::pointer CartesianPowerIterator<ForwardIt>::operator->() const
+{
+    return &getValues();
+}
+
+template<typename ForwardIt>
+inline const typename ForwardIt::value_type & CartesianPowerIterator<ForwardIt>::operator[](size_t n) const
+{
+    return getValue(n);
+}
+
+template<typename ForwardIt>
+inline typename CartesianPowerIterator<ForwardIt>::reference CartesianPowerIterator<ForwardIt>::getValues() const
+{
+    return mValues;
+}
+
+template<typename ForwardIt>
+inline const typename ForwardIt::value_type& CartesianPowerIterator<ForwardIt>::getValue(size_t n) const
```

Comment:
If we just have one, I would prefer to have `getValue`, as its meaning is clearer. `operator[]` only indicates its meaning from its return type.